### ISSUE TITLE
Accessing service IPs from Kubernetes master

### ISSUE THREAD

**Issue by lanoxx** [ovn-org/ovn-kubernetes#756]
I am using the Ansible scripts to setup a Kubernetes cluster. It seems native services such as the API server that run on the master node are not able to access ips on the overlay network such as service ips (e.g. https://10.96.0.164:443). Is this intentional and is there anyway I can change the configuration, such that the API server can access service IPs from the overlay network?

May current problem is like this. I run cert-manager in the cluster, which expects the API server to call the service IP of cert-manager-webhook pod, but of course the service IP is not accessible from the master node, and so I am getting errors in the API servers log that it cannot reach the webhook endpoint.

**Comment by girishmg** [comment]
Can you try this workaround on the K8s Master node?

```
iptables -t nat -I OUTPUT -p tcp -d <k8s_svc_cluster_ip> --dport 443 -j DNAT --to-destination k8s_server_endpoint_ip:6443
```

**Comment by lanoxx** [comment]
Hi @girishmg,

I applied the `iptables` rule on my master node but it did not help. If I understand this rule correctly, then the command adds a NAT rule from the service ip (10.96.0.164) to the pod ip (10.245.0.12). This would require, that pod IPs are accessible from the master, which is not the case.

Should pod IPs be routable on the master node? What do I need to do to achieve this?

**Comment by girishmg** [comment]
@lanoxx  by `k8s_server_endpoint_ip:6443`, I meant k8s API server IP address. Sorry, about the confusion.

**Comment by girishmg** [comment]
@lanoxx Can you please share the output of `kubectl describe svc --all-namespaces`?

**Comment by lanoxx** [comment]
@girishmg Will post it in a moment, do you happen to be on IRC in the #openvswitch channel?

**Comment by lanoxx** [comment]
Hi @girishmg here is the output of `kubectl get svc --all-namespaces`. You can find the details here in the following gist: https://gist.github.com/lanoxx/b8c28c46966e6efc6b256fe86f050b1d

    NAMESPACE                 NAME                                               TYPE           CLUSTER-IP    EXTERNAL-IP                           PORT(S)
    cert-manager              cert-manager-webhook                               ClusterIP      10.96.0.164   <none>                                443/TCP
    default                   cm-acme-http-solver-7spt2                          NodePort       10.96.0.17    <none>                                8089:30743/TCP
    default                   kubernetes                                         ClusterIP      10.96.0.1     <none>                                443/TCP
    default                   win-webserver                                      NodePort       10.96.0.129   <none>                                80:31623/TCP
    kube-system               dashboard-kubernetes-dashboard                     ClusterIP      10.96.0.92    <none>                                443/TCP
    kube-system               kube-dns                                           ClusterIP      10.96.0.10    <none>                                53/UDP,53/TCP,9153/TCP
    kube-system               nginx-ingress-controller                           LoadBalancer   10.96.0.66    XXXX.eu-central-1.elb.amazonaws.com   80:32480/TCP,443:30022/TCP
    kube-system               nginx-ingress-default-backend                      ClusterIP      10.96.0.189   <none>                                80/TCP
    kube-system               oidc-proxy-dashboard                               NodePort       10.96.0.101   <none>                                443:30754/TCP
    kube-system               tiller-deploy                                      ClusterIP      10.96.0.11    <none>                                44134/TCP

**Comment by lanoxx** [comment]
P.S. I have removed several internal services from that output, but all the essential ones should be there.

**Comment by lanoxx** [comment]
@girishmg 

>by k8s_server_endpoint_ip:6443, I meant k8s API server IP address. Sorry, about the confusion.

I don't understand what that will do. I am adding this rule on the master node where the api server runs, so I would use `127.0.0.1:443` (my k8s api server runs on `443`). But I don't understand how that would help the API server to reach the `webhook` pod of the `cert-manager` which runs on a worker node. It seems that if I add that rule, then I would create a loop where requests by the api-server end up at the api server again.

Here is an output of the API server log with the error that I am trying to solve:

> Jul 12 15:01:12 ip-172-34-44-26 kube-apiserver[9828]: E0712 15:01:12.269910    9828 available_controller.go:311]  v1beta1.admission.certmanager.k8s.io failed with: Get https://10.96.0.164:443: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)

**Comment by lanoxx** [comment]
Some more information, there is currently not even a route on the K8s master that points to the pod and svc overlay networks. When I run the following on my master node I always get empty responses:

    root@ip-172-34-44-26:/home/ubuntu# ip ro |grep 10.245 # get pod overlay route
    root@ip-172-34-44-26:/home/ubuntu# ip ro |grep 10.96  # get svc overlay route
    root@ip-172-34-44-26:/home/ubuntu#

When I run the same on a worker node I do get routes however:

    ubuntu@ip-172-34-37-91:~$ ip ro |grep 10.245
    10.96.0.0/24 via 10.245.2.1 dev k8s-ip-172-34-3
    10.245.0.0/16 via 10.245.2.1 dev k8s-ip-172-34-3
    10.245.2.0/24 dev k8s-ip-172-34-3  proto kernel  scope link  src 10.245.2.2
    ubuntu@ip-172-34-37-91:~$ ip ro |grep 10.96
    10.96.0.0/24 via 10.245.2.1 dev k8s-ip-172-34-3
    ubuntu@ip-172-34-37-91:~$

**Comment by girishmg** [comment]
@lanoxx I am on #openvswtich as `gmg`. I am there right now.

**Comment by lanoxx** [comment]
I am there now as well.

**Comment by lanoxx** [comment]
It seems that the Ansible scripts provided by ovn-kubernetes are not setting up a kubelet and overlay network on the master node itself. In particular, for the master, the `minion` group_var [defaults to false](https://github.com/ovn-org/ovn-kubernetes/blob/master/contrib/inventory/group_vars/kube-master#L6).

This is why there is no overlay network on the master. Certain software such as `cert-manager` however, assume, that the API server on the master node is able to access the overlay network IPs such as the service network.

**Comment by shettyg** [comment]
Generally, windows should move to OVS 2.11.x as there are a couple of features there that ovn-kubernetes uses. Tip of ovn-kubernetes will likely not work with OVS 2.10.x

**Comment by lanoxx** [comment]
@shettyg With ovn-kubernetes master and OVS 2.10 I was able to setup a working master which also acts as minion, though I had to make some manual adjustments because the ansible scripts dont fully support this use case. 

However I currently cannot get my windows nodes to function in this cluster. Issue #696 has a discussion about that problem. Do you think that my problems with Windows 1809 might be resolved if I switch to OVS 2.11?

**Comment by lanoxx** [comment]
I am going to close this issues as the problem can be fixed by installing `kubelet` and `ovnkube` on the master node. The Ansible scripts provided in `ovn-kubernets/contrib/` do not do that by default, but its possible, see my earlier comment: https://github.com/ovn-org/ovn-kubernetes/issues/756#issuecomment-512143131

The comments about windows in the earlier two comments are unrelated to this issue.
